src/Compilers/Core/Portable/Compilation/CompilationOptions.cs

```diff
@@ -652,13 +686,15 @@ protected int GetHashCodeHelper()
                    Hash.Combine(this.ReportSuppressedDiagnostics,
                    Hash.Combine(this.ScriptClassName != null ? StringComparer.Ordinal.GetHashCode(this.ScriptClassName) : 0,
                    Hash.Combine(Hash.CombineValues(this.SpecificDiagnosticOptions),
+                   Hash.Combine(Hash.CombineValues(PerTreeDiagnosticOptions.Keys),
+                   Hash.Combine(Hash.CombineValues(PerTreeDiagnosticOptions.Values),
```

Comment:
will this work properly with the immutable dictionaries that are values?

Is there a reason we couldn't just do Hash.CombineValues(PerTreeDiagnosticOptions) ?